target/linux/ar71xx/files/arch/mips/ath79/routerboot.c

```diff
@@ -98,20 +279,15 @@ routerboot_find_tag(u8 *buf, unsigned int buflen, u16 tag_id,
 	}
 
 	ret = -ENOENT;
-	while (buflen > 2) {
-		u16 id;
-		u16 len;
-
-		len = get_u16(buf);
-		buf += 2;
-		buflen -= 2;
-
-		if (buflen < 2)
-			break;
+	while (buflen > 4) {
+		u32 id_and_len = get_u32(buf);
+		buf += 4;
+		buflen -= 4;
+		id = id_and_len & 0xFFFF;
+		len = id_and_len >> 16;
 
-		id = get_u16(buf);
-		buf += 2;
-		buflen -= 2;
+		if (align)
+			len += (4 - len % 4) % 4;
```

Comment:
This is wrong. Use ALIGN() or the previous (less ugly) way of computing alignment.